src/language/term/Equality.re

```diff
@@ -0,0 +1,677 @@
+                c2,
+              )
+        )
+      | None => false
+      }
+    | (FixF(_, e, _), _) when ignore_fixpoints => exp'(e, e2)
+    | (_, FixF(_, e, _)) when ignore_fixpoints => exp'(e1, e)
+    | (FixF(_, _, _), _) => false
+    | (Fun(p1, e1, t1, f1), Fun(p2, e2, t2, f2)) =>
+      switch (pat'(p1, p2)) {
+      | Some(alphas_exp') =>
+        exp(Alphas.combine(alphas_exp', alphas_exp), alphas_typ, e1, e2)
+        && (ignore_function_types || Option.equal(typ', t1, t2))
+        && (ignore_function_names || f1 == f2)
+      | None => false
+      }
+    | (Fun(_), _) => false
+    | (Let(p1, e1, e2), Let(p2, e3, e4)) =>
+      switch (pat'(p1, p2)) {
+      | Some(alphas_exp') =>
+        exp(Alphas.combine(alphas_exp', alphas_exp), alphas_typ, e1, e3)
+        && exp(alphas_exp, alphas_typ, e2, e4)
+      | None => false
+      }
+    | (Let(_, _, _), _) => false
+
+    // Forms with type binders
+    | (TypFun(tp1, e1, _), TypFun(tp2, e2, _)) =>
+      switch (tpat(tp1, tp2)) {
+      | Some(alphas_typ') =>
+        exp(alphas_exp, Alphas.combine(alphas_typ', alphas_typ), e1, e2)
+      | None => false
+      }
+    | (TypFun(_, _, _), _) => false
+    | (TyAlias(tp1, t1, e1), TyAlias(tp2, t2, e2)) =>
+      switch (tpat(tp1, tp2)) {
+      | Some(alphas_typ') =>
+        typ'(t1, t2)
+        && exp(alphas_exp, Alphas.combine(alphas_typ', alphas_typ), e1, e2)
+      | None => false
+      }
+    | (TyAlias(_, _, _), _) => false
+
+    // Forms with environments. (Note fix also has an environment and is handled above.)
+    | (Closure(env1, e1), Closure(env2, e2)) when closures_by_id =>
+      ClosureEnvironment.id_equal(env1, env2) && exp'(e1, e2)
+    | (Closure(_, _), Closure(_, _)) =>
+      failwith("full closure equality has not been implemented yet")
+    | (Closure(_), _) => false
+
+    // Constructors: might ignore constructor types.
+    | (Constructor(c1, _), Constructor(c2, _))
+        when ignore_constructor_types == true =>
+      c1 == c2
+    | (Constructor(c1, Some(Some(ty1))), Constructor(c2, Some(Some(ty2)))) =>
+      c1 == c2 && typ'(ty1, ty2)
+    | (Constructor(c1, Some(None)), Constructor(c2, Some(None)))
+    | (Constructor(c1, None), Constructor(c2, None)) => c1 == c2
+    | (Constructor(_), _) => false
+
+    // Other forms: compare.
+    | (Invalid(s1), Invalid(s2)) => s1 == s2
+    | (Invalid(_), _) => false
+    | (EmptyHole, EmptyHole) => true
+    | (EmptyHole, _) => false
+    | (Undefined, Undefined) => true
+    | (Undefined, _) => false
+    | (Deferral(pos1), Deferral(pos2)) => pos1 == pos2
+    | (Deferral(_), _) => false
+    | (Atom(c1), Atom(c2)) => c1 == c2
+    | (Atom(_), _) => false
+    | (Label(l1), Label(l2)) => l1 == l2
+    | (Label(_), _) => false
+    | (LivelitName(s1), LivelitName(s2)) => s1 == s2
+    | (LivelitName(_), _) => false
+    | (MultiHole(xs1), MultiHole(xs2))
+        when List.length(xs1) == List.length(xs2) =>
+      List.equal(any', xs1, xs2)
```

Comment:
Coming across this today. Is this not a form of "provenance" for the hole? We seem to ignore this at the type level but not the expression.